Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs

```diff
@@ -1729,16 +1729,33 @@ private int GetPointerStateIndexFor(InputControl control, bool createIfNotExists
             // Determine the pointer (and touch) ID. We default the pointer ID to the device
             // ID of the InputDevice.
             var controlParent = control.parent;
+            // NOTE: m_PointerTouchControls stores references to TouchControl instances. e.g. this means that "touch0"
+            // in the same frame can have different touchId values.
             var touchControlIndex = m_PointerTouchControls.IndexOfReference(controlParent);
             if (touchControlIndex != -1)
             {
-                // For touches, we cache a reference to the control of a pointer so that we don't
-                // have to continuously do ReadValue() on the touch ID control.
-                m_CurrentPointerId = m_PointerIds[touchControlIndex];
-                m_CurrentPointerIndex = touchControlIndex;
-                m_CurrentPointerType = UIPointerType.Touch;
-
-                return touchControlIndex;
+                // Pointers are deallocated in the frame after their release (unpressed), which means the pointer state
+                // and the m_PointerTouchControls entry is only removed in the next frame
+                // (see UI_TouchPointersAreKeptForOneFrameAfterRelease).
+                // To accomodate for cases when touch is released and pressed in the same frame, the pointer
+                // state touchId needs to be checked against the cached touch control touchId (in m_PointerTouchControls)
+                // This is because the pointer state of the released touch (unpressed) still exists.
+                // If touchIDs are different, a new pointer should be allocated. Otherwise, a cached reference will
+                // be used.
+                var pointerTouchControl = m_PointerTouchControls[touchControlIndex];
+                ref var pointerState = ref GetPointerStateForIndex(touchControlIndex);
+
+                if (!(pointerTouchControl is TouchControl) ||
+                    (pointerTouchControl is TouchControl &&
+                     pointerState.eventData.touchId == ((TouchControl)pointerTouchControl).touchId.value))
```

Comment:
should we search in m_PointerTouchControls the touchid to prevent creating a new one?

Reply:
I think I didn't understand what you mean. Do you mean prevent creating a `pointerTouchControl` ? 
We also don't create touchIDs here, the native side provides them.
Not sure if any of this helps but feel free to clarify.

Reply:
if the new constrain doesn't match, the function will call `AllocatePointer` for a new pointer with the same `control` but it will be never accessible since the previous one will always be hit before.
You might tune/add a new function for `m_PointerTouchControls.IndexOfReference(controlParent);` to search for the right one.

Reply:
>  the function will call AllocatePointer for a new pointer with the same control but it will be never accessible since the previous one will always be hit before.

This is what we want: to allocate a new pointer when the touchId on the control and in the `pointerState.eventData` are different. 
An example: When we release and press a touch in the same frame this happens:
- Release: m_PointerTouchControls[0] touchId = 1, pointerState.eventData.touchId = 1;
- Press: m_PointerTouchControls[0] touchId = 2, pointerState.evenData.touchId = 1;

Ideally, this check should return -1 ` m_PointerTouchControls.IndexOfReference(controlParent);` but it doesn't since we're not storing copies of InputControl objects (it seems).
The reason seems to be that the control reference in m_PointerTouchControls[0] is always referencing the "touch0" control which always seems to have [the same position in memory](https://github.com/Unity-Technologies/InputSystem/blob/2125c295976ad25cdc69bd67cd4276936fd13510/Packages/com.unity.inputsystem/InputSystem/Devices/Touchscreen.cs#L523) and its content gets updated.

If we don't allocate a new pointer, basically we will assume there's no other pointer and that we're still "updating" the one with touchId = 1 which is not correct and generates the bug this PR is fixing.

Allocating a new pointer will create a new entry in `m_PointerTouchControls` with the same touchControl for 1 frame, as when we call `RemovePointerAtIndex` in the next frame, the entry for touchId = 1 will be removed.

Reply:
If you change a lit bit the test with a third finger on the second button you will have some remaining pointers and at.
```
`       // Release left object and Touch right object on the same frame.
        var secondPosition = scene.From640x480ToScreen(350, 200);
        EndTouch(1, firstPosition);
        BeginTouch(2, secondPosition);
        BeginTouch(3, secondPosition);
        MoveTouch(2, secondPosition);`
```
The interleave of 2/3/2 is important to create extra pointer not 'saved' by
https://github.com/Unity-Technologies/InputSystem/pull/2033/files#diff-3b5d28ba317b2b6240ae00166a94786980b2126bde8a6d830b6c962581b44583R1786 
```
        // End second touch.
EndTouch(2, secondPosition);
EndTouch(3, secondPosition);
yield return null;
```

Reply:
Now I understand the impact, thanks!

<img width="861" alt="image" src="https://github.com/user-attachments/assets/f41fdb22-9803-493c-aa49-6b5cde0dd2e6">

On the MoveTouch(2),  `m_PointerTouchControls.IndexOfReference(controlParent);`  will return index 0, which has the wrong PointerId later on, and will allocate an irrelevant pointer. Good catch 👍 
I'll fix this

Reply:
Fixed